### ISSUE TITLE
Assertion: Shared futures and unique futures can satisfy the same concept

### ISSUE THREAD

**Issue by brycelelbach** [executors/futures#27]
I'm now pretty confident that our concept tree can look like this:

```
         SemiFuture
             |
     ContinuableFuture
      |             |
UniqueFuture   SharedFuture
```

Where `UniqueFuture` and `SharedFuture` have the same API as `ContinuableFuture`, but specify some additional requirements/restrictions on those APIs. For example:

* `UniqueFuture`'s `.then` can only be called once (see #25 if you want to fight me on this).
* `SharedFuture` continuations can take arguments by value or const lvalue reference.
* `UniqueFuture` continuations can take arguments by value, const lvalue reference or rvalue reference.

**Comment by brycelelbach** [comment]
@dhollman please review.

**Comment by LeeHowes** [comment]
From a fundamental point of view, you can probably remove UniqueFuture here. Instead add a requirement in terms of permissions on SharedFuture: specifically that SharedFuture adds the requirement that via, then and other continuation-like functions may be called more than once. It also adds the requirement that the future be copyable, I think.

The question is whether the extra r-value ref support is important enough.

**Comment by brycelelbach** [comment]
Good point Lee. At first I was thinking that SharedFuture might be able to be eliminated, because UniqueFuture just seemed like SharedFuture + restrictions. Then I realized that SharedFuture also had a restriction (the continuation shouldn't take an rvalue reference parameter). I didn't think about whether UniqueFuture could be eliminated.

There is also the matter of Copyability though. The SharedFuture concept should require it, but none of the more fundamental requirements can.

I don't think we can treat SharedFuture as a strict refinement of UniqueFuture or make ContinuableFuture == UniqueFuture and make SharedFuture a refinement of that. SharedFuture doesn't just add requirements, it changes them - e.g. how many times .then can be called. We might want to consider having UniqueFuture::then be rvalue-reference qualified (although we explored this in P0701 and I think it's a bad idea) which would add another difference. I'll open a separate issue for that.

**Comment by dhollman** [comment]
> @dhollman please review.

Looks great to me.

> The question is whether the extra r-value ref support is important enough.

As Bryce said, Copyability is a far bigger issue IMHO. It can be hard to write code that's generic across copy and non-copy types, so it's useful to separate the concepts.

I also think there's some value in leaving `ContinuableFuture` as the bare-bones thing that executors require and having another concept for the user-friendly interface. For instance, it allows us to put out adapters that only require what executors need and provide the friendly `then` with all the bells and whistles (e.g., unwrapping, etc.) — to do this, you need to specify a minimum concept for the type that the adapter operates on.

**Comment by brycelelbach** [comment]
I'd propose that we go forward with the concept tree I suggest at the start of the thread.

Proposed action items:

* Add refined concepts `SharedFuture` and `UniqueFuture`

**Comment by brycelelbach** [comment]
As per #26, we do want to support non-const lvalue references in shared future continuations.

With that, we have:

* `SharedFuture` continuations can take arguments by value, const lvalue reference or non-const lvalue reference.
* `UniqueFuture` continuations can take arguments by value, const lvalue reference or rvalue reference

**Comment by LeeHowes** [comment]
> As Bryce said, Copyability is a far bigger issue IMHO. It can be hard to write code that's generic across copy and non-copy types, so it's useful to separate the concepts.

But it's easy to write code that assumes non-copyability, and add a concept requirement to add copyability. That is a filtering of the valid set of types by adding requirements to them.

Incidentally, filtering out && isn't enough if there are any other places where we might allow a move-only T as a parameter. That wouldn't work on a shared-future either. 


> I also think there's some value in leaving ContinuableFuture as the bare-bones thing that executors require and having another concept for the user-friendly interface. 

I agree, but it doesn't change this I think because the fundamental unique_future/shared_future difference applies to .continue_with. 

I think the only difference here is that shared_future is a continuable_future that: 
 * That is copy constructible
 * Requires a copy-constructible T
 * And if we r-value qualify functions, that supports an l-value qualified continue_with operation.

**Comment by LeeHowes** [comment]
And my comment in #26 was:

So I think my position on this is that we should always support r and l-value parameters, and it is up to the call wrapper to pass a reference to a temporary if that is the requirement, or to move the temporary in after a copy for an r-value version off a shared future.

I think we may require that shared_future's T be copyable, because anything else is going to mean some very T-specific and maybe implementation-specific rules on the continuation parameters.

**Comment by jaredhoberock** [comment]
> I think we may require that shared_future's T be copyable, because anything else is going to mean some very T-specific and maybe implementation-specific rules on the continuation parameters.

What is the reasoning behind this requirement? It seems conceptually similar to requiring that `shared_ptr`'s `T` be copyable.

**Comment by LeeHowes** [comment]
We just have to be careful to not be too confusing about what we allow as continuation parameters.

For a unique_future we can always move. So T, T&, const T& and T&& are all valid.

For a shared_future we can do the same if we allow copies. If we do not allow copies we can only  support T& and const T&.
We could define the rule that T& and const T& are always supported. T and T&& are supported if T is copyable (for a shared_future) and movable (T&& on a unique_future).

It is the same as shared_ptr in that operator* does only return the reference forms, everything else is a check on T's capabilities. Here we want to make sure the code gets a consistent compile-time error if we can, or it will depend on how the implementation implements the shared_future.

**Comment by LeeHowes** [comment]
Incidentally, I should point out that passing ready_future by value to the continuation as default in continue_with, allows the implementation to hide this under the ready_future parameter, and ready_futures getters are ref-qualified.

As does suggesting that we don't specify the valid forms of the value parameter and suggest that people use auto&& to be generic on an unknown future type.

**Comment by dhollman** [comment]
> For a unique_future we can always move. So T, T&, const T& and T&& are all valid.

Wait, how is it valid to move if the parameter is `T&`? Supporting that (relatively corner) case requires a lot more implementation effort, IMHO.

> For a shared_future we can do the same if we allow copies. If we do not allow copies we can only support T& and const T&.

You could vary the supported signatures based on the copyability of `T`. If the user choses to write code that's not generic with respect to copy constructibility of `T`, then they should get errors when given a non-copyable `T`; I don't see anything wrong with that.

> As does suggesting that we don't specify the valid forms of the value parameter and suggest that people use auto&& to be generic on an unknown future type.

I still think you have to say things about whether or not the parameter may be a copy. Otherwise people can try to do things that are dependent on the object identity of the reference parameter (e.g., take its address and do something with it), and we need to say something explicit to prohibit this in generic code, just like we do for the predicates in the algorithm header.

**Comment by LeeHowes** [comment]
Also answering David's point in #26 here so we can really kill the closed issue. 

Maybe you're right. T&& is ok for unique because we still get reference semantics. It is slightly clearer for the lifetime than an l-value ref. The l-value ref works fine and isn't hard to implement, the lifetime of the temporary outlives the function call. It's just we need out of band description to make it clear that that reference does not live past the end of the callback, so cannot be stored.

The reason I was suggesting we want to allow copy is that in the folly implementation we have found that require copies for shared_promise semantics is a clean design. We implement shared promise as a chain of promises and the shared promise completes by move in, store, copy + move into each of the promises in the chain, and destroy. I'm slightly wary of disallowing that design, and it certainly does not need reference counting.

The reason this is simpler is that shared_promise.get_future() returns a unique_future (and of course shared_future can use a shared_promise internally as its means of implementation). That unique_future knows no difference in terms of the implementation. 

If we do expect to reference count, then we need to handle the unpacking from the reference counting state somewhere. I think that is probably fine because shared_future.then wraps the callback in something that handles that, passes a reference to the contents, and it is up to the callback how the move-only but non-movable type is converted to a useful return value. 

It's not the design I'm used to so it isn't natural, but I don't have a strong argument against it.


So assuming that we do go with this plan of T& and const T& for shared_future, T and T&& for unique_future. How do we distinguish between shared_future and unique_future in the concept? Should it be that shared_future has l-value-qualified .then + a copy constructor and unique_future has r-value-qualified .then?

**Comment by brycelelbach** [comment]
> And my comment in #26 was:
>
> So I think my position on this is that we should always support r and l-value parameters, and it is up to the call wrapper to pass a reference to a temporary if that is the requirement, or to move the temporary in after a copy for an r-value version off a shared future.

I think it's a really high burden to expect unique future authors to make non-constant l-value references work - they'd have to introspect the callable to figure out what it can be called with. Also, semantically it doesn't make sense. If I write: `unique_future<int> f; f.then([] (int& i) { /* ... */});`, any changes I make to `i` are going to be meaningless; no one else will ever see them. It's not meaningful, so we shouldn't go out of our way to support it or disallow it. Unique futures should be required to work with continuations can take arguments by value, const lvalue reference or rvalue reference; and no more than that.

**Comment by brycelelbach** [comment]
Sorry, perhaps I was not clear earlier about **why** the continuation signatures I suggested should work and why others should not.

`.then` on a `UniqueFuture` conceptually moves the value out of the shared state and passes it to the continuation. E.g. given a continuation `f`, a `UniqueFuture` does the equivalent of `f(std::move(shared_state->value))`.

`.then` on a `SharedFuture` conceptually passes a non-const lvalue reference to the continuation. E.g. given a continuation `f`, a `UniqueFuture` does the equivalent of `f(shared_state->value)`.

Some have suggested that we should go out of our way to have `UniqueFuture` support continuations that take their argument by non-const. Why? This makes implementors of `UniqueFuture` go out of their way to do something quirky, and is not how C++ generally works.

We shouldn't be thinking about what continuation signatures we support. Instead, we should look at what type of reference `UniqueFuture` and `SharedFuture` pass to the continuation; the types of continuations supported will be a consequence of that.

The rule is simple:

* **`UniqueFuture`s call their continuation with an rvalue reference to the value.**
* **`SharedFuture`s call their continuation with a non-constant lvalue reference to the value.**

What does this mean?

For `UniqueFuture`:

```
void f(std::string s)        { std::cout << s << std::endl; }
void g(std::string& s)       { std::cout << s << std::endl; }
void h(std::string const& s) { std::cout << s << std::endl; }
void i(std::string&& s)      { std::cout << s << std::endl; }

int main()
{
  std::string s0 = "hello";
  std::string s1 = "from";
  std::string s2 = "planet";
  std::string s3 = "earth";

  f(std::move(s0));
  g(std::move(s1)); // COMPILATION FAILURE
  h(std::move(s2));
  i(std::move(s3));
}
```

`UniqueFuture` continuations can take their argument by value, by non-const lvalue reference, or by rvalue reference.

For `SharedFuture`:

```
void f(std::string s)        { std::cout << s << std::endl; }
void g(std::string& s)       { std::cout << s << std::endl; }
void h(std::string const& s) { std::cout << s << std::endl; }
void i(std::string&& s)      { std::cout << s << std::endl; }

int main()
{
  std::string s0 = "hello";
  std::string s1 = "from";
  std::string s2 = "planet";
  std::string s3 = "earth";

  f(s0);
  g(s1);
  h(s2);
  i(s3); // COMPILATION FAILURE
}
```

`SharedFuture` continuations can take their arguments by value, by non-const lvalue reference, or by rvalue reference.

**Comment by brycelelbach** [comment]
So the fundamental question here is whether `SharedFuture`s are a refinement of `UniqueFuture`.

E.g. Is a `SharedFuture` also a `UniqueFuture`?

As an analog, think about iterators. All `RandomAccessIterator`s are `InputIterator`s and fulfill all the requirements of `InputIterator`.

It seems to me that `UniqueFuture::then` and `SharedFuture::then` are different operations. One consumes, and one observes. Perhaps we are trying to hard to make them work.

Let's assume that the right answer to #26 is that `SharedFuture` cannot modify its state. That is, we tweak my two rules from above, saying that `SharedFuture`s call their continuations with constant lvalue references, not non-constant lvalue references:

* `UniqueFuture`s call their continuation with a rvalue reference to the value.
* `SharedFuture`s call their continuation with a constant lvalue reference to the value.

We still have a problem - `SharedFuture` is still not a strict refinement of `UniqueFuture`, because `UniqueFuture` continuations can be called with a rvalue reference and `SharedFuture` continuations cannot.

So, I think we **do** have four concepts:

* `SemiFuture`.
* `ContinuableFuture` is `Movable`, requires `T` to be `Movable`, has `.then` which can be called no more than once, and supports continuations that take their parameter by value or by const lvalue reference.
* `SharedFuture` refines `ContinuableFuture`. It is `Copyable`, requires `T` to be `Copyable`, has `.then` which can be called more than once, and supports continuations that take their parameter by value or by const lvalue reference.
* `UniqueFuture` refines `ContinuableFuture`. It supports continuations that take their parameter by rvalue reference.

`UniqueFuture` is a very subtle refinement of `ContinuableFuture` - all it gives you is the guarantee that it can be called with a continuation that binds to rvalue references. 

When would you need this? Only when you have types that are expensive to move, such as big POD types.

So, should we have a `UniqueFuture` concept? I think the answer is **yes**, just for completeness. But it doesn't have to be in version 1, necessarily.

Note that this is predicated on the assumption that our answer to #26 is that we don't need to modify the state of a `SharedFuture`.

**Comment by griwes** [comment]
Non-movable, non-copyable types seem so rare to me that I think we could just require movability. (Plus, as somebody mentioned before, you can turn a non-movable type into a movable type by adding a layer of indirection, i.e. having a future to an `unique_ptr<T>` or something like that, instead of `T`.)

**Comment by LeeHowes** [comment]
This whole thread is making me very doubtful that shared_future is a clean model to have here. I better understand why folly did not go with a shared_future design at all and why every time I mention it people grimace.

Should we consider making a fundamentally different construct for sharing of futures? A type that is not really a future at all, but is copyable, chains itself onto an input future and can be make into a `UniqueFuture<reference_wrapper<T>>`. That would be more consistent.

So we'd do something like:
```
Future<int> f;
auto sf = split_future(std::move(f));
auto sf2 = sf;
sf.toFuture().then([](reference_wrapper<int> r){...});
sf.toFuture().then([](reference_wrapper<int> r){...});
```

sf can be a SemiFuture, but not a Future, in that for consistency reasons we make the to/from sharing transitions completely obvious. Otherwise if we receive a "Future" as a parameter to a generic function the behaviour of .then depends on whether it is shared or not!

**Comment by LeeHowes** [comment]
Or, at the very least, because we have explicitly wrapped the parameter in a reference wrapper, then it really is just a future, but one with a copy constructor. Such that SharedFuture is a superset of Future, it's just that to get one you do a minor wrapping of the future value type so that there is no confusion.

**Comment by griwes** [comment]
I don't know what I think about that. I like Bryce's last list of concepts, and shared futures are useful enough for an entire class of problems (basically any computational DAGs) that I think there's something fundamental there. I don't really think that those four concepts are problematic in any way.

We know how to implement shared futures. I don't think we can do that lockfree, but we can definitely do that with locks - it's not too horrible, and the value is hard to overestimate IMO.

Now, for the generic code argument, I think Bryce's definition perfectly captures what a `ContinuableFuture` is. The concept guarantees that you can call `.then` once, with something that can be called with a const lvalue. This is already enough information to be able to implement a lot of algorithms.

And since by-value arguments are allowed there, then you can write some of the algorithms the same way for shared and for unique, and get the "unique future moves its value" thing kinda automatically gives you an optimization. As long as you stick to just using what the concept gives you, I don't see a problem with the existence of `SharedFuture` there.

Of course any other behavior is not defined, but there's a lot of other parts of the semantics that are going to differ regardless of what concept you take as a parameter.

**Comment by LeeHowes** [comment]
By-value generic code would have to allow copyable T to work with SharedFuture. So we'd have to be a little careful writing such code.

**Comment by dhollman** [comment]
> By-value generic code would have to allow copyable T to work with SharedFuture. So we'd have to be a little careful writing such code.

But you can just do the same thing with `const T&` arguments instead and get the same genericness.

**Comment by LeeHowes** [comment]
Not entirely, because when you pass a const T& you need to make the argument copyable to do much with it anyway. 

Either way, I'm happy with what we got to on Thursday, so there is no more debate necessary here.

**Comment by brycelelbach** [comment]
We discussed this at the NVIDIA F2F, and decided we want three concepts for version 1:

```
SemiFuture
ContinuableFuture
SharedFuture
```

Action Item:
* [ ] Write/adjust wording for these three concepts.
* [ ] Write/adjust wording for continuation requirements.

**Comment by LeeHowes** [comment]
Poll from F2f in #44
